lib/active_merchant/billing/gateways/cyber_source_rest.rb

```diff
@@ -245,68 +305,95 @@ def add_merchant_description(post, options)
         URI.parse(url('')).host
       end
 
       def parse(body)
+        return {} if body.blank?
+
         JSON.parse(body)
       end
 
-      def commit(action, post)
-        response = parse(ssl_post(url(action), post.to_json, auth_headers(action, post)))
+      def commit(action, post, http_method = :post)
+        response = parse(ssl_request(http_method, url(action), post.nil? || post.empty? ? nil : post.to_json, auth_headers(action, post, http_method)))
+        succeeded = success_from(action, response, http_method)
+        body = action == :delete ? { response_code: response.to_s } : response
         Response.new(
-          success_from(response),
-          message_from(response),
+          succeeded,
+          message_from(body, succeeded, http_method),
           response,
-          authorization: authorization_from(response),
+          authorization: authorization_from(response, action),
           avs_result: AVSResult.new(code: response.dig('processorInformation', 'avs', 'code')),
           # cvv_result: CVVResult.new(response['some_cvv_response_key']),
           test: test?,
-          error_code: error_code_from(response)
+          error_code: error_code_from(response, succeeded)
         )
       rescue ActiveMerchant::ResponseError => e
         response = e.response.body.present? ? parse(e.response.body) : { 'response' => { 'rmsg' => e.response.msg } }
         Response.new(false, response.dig('response', 'rmsg'), response, test: test?)
       end
 
-      def success_from(response)
-        %w(AUTHORIZED PENDING REVERSED).include?(response['status'])
+      def success_from(action, response, http_method)
+        return response.empty? if http_method == :delete
+
+        case action
+        when /payments/
+          %w(AUTHORIZED PENDING REVERSED).include?(response['status'])
+        else
+          return response['id'].present? && !response['errorInformation'].present? unless http_method == :delete
+        end
       end
 
-      def message_from(response)
-        return response['status'] if success_from(response)
+      def message_from(response, succeeded, http_method)
+        return 'OK' if succeeded && http_method == :delete
+        return response['status'] if succeeded
 
         response['errorInformation']['message']
       end
 
-      def authorization_from(response)
+      def authorization_from(response, action)
         id = response['id']
         has_amount = response['orderInformation'] && response['orderInformation']['amountDetails'] && response['orderInformation']['amountDetails']['authorizedAmount']
         amount = response['orderInformation']['amountDetails']['authorizedAmount'].delete('.') if has_amount
-
+        if /payment-instruments/.match(action) && !response.empty?
+          customer_id = response['_links']['customer']['href'].split('/').last
+          payment_instrument_id = response['id']
+          return [customer_id, payment_instrument_id].join('|')
```

Comment:
I think we should keep the ordering of elements in an authorization string as consistent as we can. To that end, there should only be one final concatenation, which, if there are any elements missing, are just blank. For example, 

[response_id, amount, customer_id, payment_instrument_id].join('|')

This lets us know for sure which element means what, for any given authorization string.

Reply:
I think we should also record the `instrument_id` from the second call, if it is available.

Reply:
Hello @curiousepic, this issue was changed, please take a look of the new version (amend on this same PR) and let me know if the change complies with the suggestion.